dynamic_programming/longest_common_subsequence.py

```diff
@@ -69,14 +72,73 @@ def longest_common_subsequence(x: str, y: str):
     return L[m][n], seq
 
 
+"""
+This is the Top-Down Approach (Memoization) of Dynamic Programming.
+"""
+
+
+def longest_common_subsequence_memoization(X, Y, m, n, dp):
+    """
+    :param X: String 1st
+    :param Y: String 2nc
+    :param m: length of String 1
+    :param n: length of String 2
+    :param dp: Array for Storage of The Pre Calculate Value
```

Comment:
If we shorten the strings before recursion, then m and n can be calculated instead of passed in.

Can dp can be memoized using functools.lru_cache()?

Reply:
Sir if shorten the m and n then it will create problem while calculating reducing Part.
**function_name(x,y,m-1,n-1,dp)**
We call the recursively and shorten the length of String.